File: C++/binary-tree-traversal.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int val;
    Node *left, *right;
    Node(int v) : val(v), left(nullptr), right(nullptr) {}
};

Node* buildTree(vector<int> &vals, int &idx) {
    if (idx >= (int)vals.size() || vals[idx] == -1) { idx++; return nullptr; }
    Node* node = new Node(vals[idx++]);
    node->left = buildTree(vals, idx);
    node->right = buildTree(vals, idx);
    return node;
}

int main() {
    vector<int> vals;
    int x;
    while (cin >> x) vals.push_back(x);

    int idx = 0;
    Node* root = buildTree(vals, idx);

    queue<Node*> q;
    if (root) q.push(root);

    while (!q.empty()) {
        int size = q.size();
        cout << "[";
        for (int i = 0; i < size; i++) {
            Node* node = q.front(); q.pop();
            cout << node->val << (i + 1 < size ? ", " : "");
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
        cout << "] ";
    }
    cout << endl;
    return 0;
}

File: JS/binary-tree-traversal.js
class Node {
    constructor(val) { this.val = val; this.left = null; this.right = null; }
}

function buildTree(vals, idxRef) {
    if (idxRef.i >= vals.length || vals[idxRef.i] === -1) { idxRef.i++; return null; }
    const node = new Node(vals[idxRef.i++]);
    node.left = buildTree(vals, idxRef);
    node.right = buildTree(vals, idxRef);
    return node;
}

function levelOrder(root) {
    const result = [];
    if (!root) return result;
    let queue = [root];

    while (queue.length) {
        const level = [];
        const next = [];
        for (const node of queue) {
            level.push(node.val);
            if (node.left) next.push(node.left);
            if (node.right) next.push(node.right);
        }
        result.push(level);
        queue = next;
    }
    return result;
}

const vals = require('fs').readFileSync('/dev/stdin', 'utf8').trim().split(/\s+/).map(Number);
const root = buildTree(vals, { i: 0 });
console.log(levelOrder(root).map(l => '[' + l.join(', ') + ']').join(' '));